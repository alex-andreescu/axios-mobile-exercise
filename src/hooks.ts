import { useEffect, useState } from 'react'
import axios from "axios";

export const useStream = () => {
  const [ data, setData ] = useState(null);
  const [ error, setError ] = useState(null);
  const [ done, setDone ] = useState(false);

  useEffect(() => {
    (async () => {
      try {
        const response = await axios.get(
          'https://api.axios.com/api/render/stream/content',
          {
            params: {
              page_size: 20,
            },
          },
        );
        setData(response.data);
      } catch (error) {
        setError(error.message);
      } finally {
        setDone(true);
      }
    })();
  }, []);

  return { data, error, done };
};

export const useContent = (uuid) => {
  const [ data, setData ] = useState(null)
  const [ error, setError ] = useState(null)
  const [ done, setDone ] = useState(false)

  useEffect(() => {
    (async () => {
      try {
        const response = await axios.get(`https://api.axios.com/api/render/content/${uuid}`);
        setData(response.data)
      } catch (error) {
        setError(error.message);
      } finally {
        setDone(true);
      }
    })();
  }, []);

  return { data, error, done };
}
