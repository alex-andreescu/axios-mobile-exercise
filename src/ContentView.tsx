import React, { useEffect, useState } from 'react';
import { Image, SafeAreaView, StyleSheet, Text, ScrollView } from 'react-native';
// I'll be honest - didn't know how to fix this error but the functionality works fine
// @ts-ignore:next-line
import getRNDraftJSBlocks from 'react-native-draftjs-render';
import { useContent } from './hooks';

const styles = StyleSheet.create({
  bold: {
    color: 'black',
    fontWeight: 'bold',
  },
  safeArea: {
    backgroundColor: 'lightgray',
    flex: 1,
    justifyContent: 'center',
  },
  scrollView: {
    paddingHorizontal: 20,
  },
  contentImage: {
    alignSelf: 'center',
    height: 300,
    width: 400,
  },
  publishedOn: {
    marginTop: 10,
  },
  author: {
    marginTop: 10,
    marginBottom: 20,
  },
})

interface ContentViewProps {
  route: {
    params: {
      uuid: string;
    },
  },
}

export const ContentView = ({ route: { params: { uuid } } }: ContentViewProps) => {
  const [ contentState, setContentState ] = useState<{}>({})
  const { data, error, done } = useContent(uuid)

  useEffect(() => {
    // only set the state if we get data back and loading is done
    if (data !== null && done && !error) {
      setContentState(data['blocks'])
    }
  }, [ data, done ])

  // I noticed some of the images were missing so i went ahead and added a default "no image available" image
  const uri: string = data?.['primary_image']?.['base_image_url'] || 'https://upload.wikimedia.org/wikipedia/commons/thumb/a/ac/No_image_available.svg/600px-No_image_available.svg.png'
  const authors: Array<string> = (data?.['authors'] || []).map(({ display_name }) => display_name)
  const authorsText: string = authors?.join(', ')
  const topics: Array<string> = (data?.['topics'] || []).map(({ name }) => name)
  const topicsText: string = topics?.join(', ')
  const publishedOn: Date = new Date(data?.['published_date'] || 0)
  const blocks = getRNDraftJSBlocks({
    contentState,
    customStyles: StyleSheet.flatten({
      paragraph: {
        color: 'black',
      },
      'ordered-list-item': {
        color: 'black',
      },
      'unordered-list-item': {
        color: 'black',
      },
      'blockquote': {
        color: 'black',
      },
      unstyled: {
        color: 'black',
      },
    }),
  })

  return (
    <SafeAreaView style={styles.safeArea}>
      <ScrollView style={styles.scrollView}>
        <Image
          style={styles.contentImage}
          source={{ uri }}
          accessible
          accessibilityLabel={data?.['primary_image']?.['alt_text']}
        />
        <Text style={[ styles.bold, styles.publishedOn ]}>{`${publishedOn.toDateString()} - ${topicsText}`}</Text>
        <Text style={[ styles.bold, styles.author ]}>{`Written by ${authorsText}`}</Text>
        {blocks}
      </ScrollView>
    </SafeAreaView>
  )
}
