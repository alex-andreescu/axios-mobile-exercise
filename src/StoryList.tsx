import React from 'react';
import { FlatList, Image, SafeAreaView, StyleSheet, Text, Pressable } from 'react-native';
import type { NativeStackNavigationProp } from '@react-navigation/native-stack';
import { useContent, useStream } from './hooks';

const styles = StyleSheet.create({
  safeArea: {
    backgroundColor: 'lightgray',
    flex: 1,
    justifyContent: 'center',
  },
  listItem: {
    justifyContent: 'center',
    flex: 1,
    flexDirection: 'row',
    marginHorizontal: 20,
    marginTop: 20,
  },
  listImage: {
    height: 105,
    width: 140,
  },
  headline: {
    flex: 1,
    fontSize: 18,
    padding: 10,
    color: 'black',
  },
  headerButton: {
    fontSize: 11,
  },
});

interface StoryListItemProps {
  navigation: NativeStackNavigationProp<any>;
  uuid: string
}

interface StoryListProps {
  navigation: NativeStackNavigationProp<any>;
}

export const StoryListItem = ({ navigation, uuid }: StoryListItemProps) => {
  const { data } = useContent(uuid)

  const onPress = () => {
    navigation.navigate('ContentView', { uuid })
  }

  // I noticed some of the images were missing so i went ahead and added a default "no image available" image
  const uri: string = data?.['primary_image']?.['base_image_url'] || 'https://upload.wikimedia.org/wikipedia/commons/thumb/a/ac/No_image_available.svg/600px-No_image_available.svg.png'

  return (
    <Pressable
      style={({ pressed }) => [
        // nice QoL for users to know that their press is happening
        { backgroundColor: pressed ? 'rgb(210, 230, 255)' : 'lightgray' },
        styles.listItem,
      ]}
      onPress={onPress}
    >
      <Image
        style={styles.listImage}
        source={{ uri }}
        accessible
        accessibilityLabel={data?.['primary_image']?.['alt_text']}
      />
      <Text style={styles.headline} allowFontScaling>{data?.['headline']}</Text>
    </Pressable>
  )
}

export const StoryList = ({ navigation }: StoryListProps) => {
  const { data } = useStream()

  const renderItem = ({ item }) => {
    return <StoryListItem uuid={item} navigation={navigation} />
  }

  return (
    <SafeAreaView style={styles.safeArea}>
      <FlatList
        data={data?.['results']}
        renderItem={renderItem}
        keyExtractor={(item) => item}
        // if no news stories are returned, let the user know via message
        ListEmptyComponent={<Text>No news stories returned</Text>}
      />
    </SafeAreaView>
  )
}
