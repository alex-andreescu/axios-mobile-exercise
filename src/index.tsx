import React from 'react';
import { NavigationContainer } from '@react-navigation/native';
import { createNativeStackNavigator } from '@react-navigation/native-stack';
import { Linking, Text, TouchableOpacity } from 'react-native';
import { ContentView } from './ContentView';
import { StoryList } from './StoryList';

const Stack = createNativeStackNavigator();

const App: React.FC = () => {
  return (
    <NavigationContainer>
      <Stack.Navigator initialRouteName="StoryList">
        <Stack.Screen
          name="StoryList"
          component={StoryList}
          options={{ headerTitle: 'Stories' }}
        />
        <Stack.Screen
          name="ContentView"
          component={ContentView}
          options={{
            headerTitle: '',
            headerRight: () => (
              <TouchableOpacity
                onPress={() => {
                  const url: string = 'https://axios.com'

                  return Linking.openURL(url)
                }}
              >
                <Text style={{ color: 'blue' }}>Visit Axios.com</Text>
              </TouchableOpacity>
            ),
          }}
        />
      </Stack.Navigator>
    </NavigationContainer>
  );
};

export default App;
