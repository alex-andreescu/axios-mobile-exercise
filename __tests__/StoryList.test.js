import React from 'react'
import ShallowRenderer from 'react-test-renderer/shallow';
import { StoryList, StoryListItem } from '../src/StoryList'

jest.mock('../src/hooks', () => ({
  useStream: jest.fn(() => ({
    data: {
      results: [ '1234', '2345', '3456' ],
    },
    error: '',
    done: true,
  })),
  useContent: jest.fn(() => ({
    data: {
      primary_image: {
        base_image_url: 'www.google.com',
        alt_text: 'google',
      },
      headline: 'a headline',
    },
    error: '',
    done: true,
  })),
}))

describe('StoryListItem', () => {
  const navigation = {
    navigate: jest.fn()
  }
  const uuid = '1234'
  const renderer = new ShallowRenderer();
  renderer.render(<StoryListItem navigation={navigation} uuid={uuid} />);
  const result = renderer.getRenderOutput();

  test('snapshot', () => {
    expect(result).toMatchSnapshot()
  })
})

describe('StoryList', () => {
  const navigation = {
    navigate: jest.fn()
  }
  const renderer = new ShallowRenderer();
  renderer.render(<StoryList navigation={navigation} />);
  const result = renderer.getRenderOutput();

  test('snapshot', () => {
    expect(result).toMatchSnapshot()
  })
})
