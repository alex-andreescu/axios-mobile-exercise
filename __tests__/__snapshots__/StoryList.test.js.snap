// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StoryList snapshot 1`] = `
<ForwardRef(SafeAreaView)
  style={
    Object {
      "backgroundColor": "lightgray",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <FlatList
    ListEmptyComponent={
      <Text>
        No news stories returned
      </Text>
    }
    data={
      Array [
        "1234",
        "2345",
        "3456",
      ]
    }
    keyExtractor={[Function]}
    renderItem={[Function]}
  />
</ForwardRef(SafeAreaView)>
`;

exports[`StoryListItem snapshot 1`] = `
<Memo(Pressable)
  onPress={[Function]}
  style={[Function]}
>
  <Image
    accessibilityLabel="google"
    accessible={true}
    source={
      Object {
        "uri": "www.google.com",
      }
    }
    style={
      Object {
        "height": 105,
        "width": 140,
      }
    }
  />
  <Text
    allowFontScaling={true}
    style={
      Object {
        "color": "black",
        "flex": 1,
        "fontSize": 18,
        "padding": 10,
      }
    }
  >
    a headline
  </Text>
</Memo(Pressable)>
`;
